Add more `MonoidK` based folds

BalmungSan opened the pull request:
I am opening this before adding tests and docs to get feedback and see if it is a worth-it addition or not; or even to confirm this doesn't exist already 😅

I found myself needing `foldSequenceK` today having a `List[Resource[IO, HttpRoutes[IO]]` and wanting to get back a `Resource[IO, HttpRoutes[IO]]` in an easy way.

BalmungSan commented on core/src/main/scala/cats/Traverse.scala:
```diff
@@ -213,6 +213,22 @@ trait Traverse[F[_]] extends Functor[F] with Foldable[F] with UnorderedTraverse[
   override def unorderedSequence[G[_]: CommutativeApplicative, A](fga: F[G[A]]): G[F[A]] =
     sequence(fga)
+
+  def foldTraverse[A, B, G[_]](fa: F[A])(f: A => G[B])(implicit G: Monad[G], M: Monoid[B]): G[B] =
+    foldM(fa, M.empty) { case (acc, a) =>
+      G.map(f(a)) { b =>
+        M.combine(acc, b)
+      }
+    }
```
I don't like needing `Monad` here, since `Applicative` should be enough using `foldLeft` and `map2`
But, it seems the implementation of `foldM` is reasonably optimized and ensures an early stop, not sure if using `map2` would achieve that as well.

What do you all think?
Any feedback is welcome.

armanbilge replied:
I think you may be able to use `map2Eval`? See also:
- https://github.com/typelevel/cats/issues/4408

armanbilge reviewed (commented):
Wait, is this `foldMapM`?

https://github.com/typelevel/cats/blob/14286c9ef89a652fcff182bbd685f6caf930701e/core/src/main/scala/cats/Foldable.scala#L545-L546

BalmungSan commented:
@armanbilge oh lol, you are right. So what was missing was only the friends?
Since I originally needed the most derived one _(`foldSequenceK`)_ I traced back and started from scratch.

Although, I should have realized it, since I first thought of calling `foldTraverseK` like `foldMapMK` but thought that name was very confusing and decided to do the parallel with `Traverse`, also I was still thinking of using `Applicative` at that moment.

armanbilge commented on core/src/main/scala/cats/Traverse.scala:
```diff
@@ -213,6 +213,22 @@ trait Traverse[F[_]] extends Functor[F] with Foldable[F] with UnorderedTraverse[
+  def foldTraverseK[A, B, G[_], H[_]](fa: F[A])(f: A => G[H[B]])(implicit G: Monad[G], M: MonoidK[H]): G[H[B]] =
+    foldTraverse(fa)(f)(G, M.algebra)
+
+  def foldSequence[A, G[_]](fga: F[G[A]])(implicit G: Monad[G], M: Monoid[A]): G[A] =
```
This is `foldA`.

https://github.com/typelevel/cats/blob/14286c9ef89a652fcff182bbd685f6caf930701e/core/src/main/scala/cats/Foldable.scala#L500-L501

BalmungSan commented:
Okay, after Arman realized I was reimplementing the well with the _"traverse"_ ones.
I just added the _"sequence"_ ones, and changed the name to be consistent... but, being totally honest, I dislike all of them, as well as the `M` vs `A` variants, those should be internal optimizations but I understand that was probably not a known idea by the time.
And I especially dislike that `foldM` takes the usual `fold` parameters instead of being like `combineAll`, but `foldMK` doesn't.
